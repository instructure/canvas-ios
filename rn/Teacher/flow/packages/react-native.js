//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

declare var console: typeof console & { disableYellowBox: boolean }

export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}

export type KeyboardEventData = {
  endCoordinates: {
    width: number,
    height: number,
    screenX: number,
    screenY: number,
  },
}
