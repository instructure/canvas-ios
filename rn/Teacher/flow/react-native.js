//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
