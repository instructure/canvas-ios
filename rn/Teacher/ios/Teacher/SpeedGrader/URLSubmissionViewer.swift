//
// This file is part of Canvas.
// Copyright (C) 2020-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

import SwiftUI
import Core

struct URLSubmissionViewer: View {
    let submission: Submission

    @Environment(\.appEnvironment) var env
    @Environment(\.viewController) var controller

    @State var image: UIImage?
    @State var loaded: URL?
    @State var error: Error?

    var body: some View {
        ScrollView {
            VStack(spacing: 0) {
                HStack {
                    Text("This submission is a URL to an external page. We've included a snapshot of what the page looked like when it was submitted.")
                        .font(.regular16).foregroundColor(.textDarkest)
                    Spacer()
                }
                Button(action: {
                    guard let url = submission.url else { return }
                    env.loginDelegate?.openExternalURL(url)
                }, label: {
                    Text(submission.url?.absoluteString ?? "")
                        .font(.semibold16).foregroundColor(Color(Brand.shared.linkColor))
                        .padding(.vertical, 16)
                })
                if let file = submission.attachments?.first, let id = file.id, let url = file.url {
                    if let error = error, url == loaded {
                        Text(error.localizedDescription)
                            .font(.regular16).foregroundColor(.textDanger)
                    } else if let image = image, url == loaded {
                        Button(action: {
                            env.router.route(
                                to: "/files/\(id)",
                                from: controller,
                                options: .modal(embedInNav: true, addDoneButton: true)
                            )
                        }, label: {
                            Image(uiImage: image)
                                .resizable()
                                .aspectRatio(image.size, contentMode: .fill)
                                .accessibility(label: Text("URL Preview Image"))
                        })
                            .padding(.horizontal, -16)
                    } else {
                        ProgressView()
                            .progressViewStyle(.indeterminateCircle())
                            .padding(32)
                            .onAppear {
                            env.api.makeRequest(url) { (data, _, error) in
                                image = data.flatMap { UIImage(data: $0) }
                                self.error = error
                                self.loaded = url
                            }
                        }
                    }
                } else {
                    Text("Preview Unavailable")
                        .font(.regular16).foregroundColor(.textDark)
                }
            }
                .multilineTextAlignment(.leading)
                .padding(16)
        }
    }
}
