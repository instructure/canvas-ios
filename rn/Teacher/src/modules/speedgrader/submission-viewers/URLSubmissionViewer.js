//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React, { Component } from 'react'
import {
  Text,
  Image,
  ScrollView,
  StyleSheet,
  Linking,
} from 'react-native'
import i18n from 'format-message'
import { Paragraph } from '../../../common/text'
import { LinkButton } from '../../../common/buttons'

type Props = {
  submission: Submission,
  drawerInset: number,
}

type State = {
  aspectRatio: number,
  size: { width: number, height: number },
}

export default class URLSubmissionViewer extends Component<Props, State> {
  state: State = {
    aspectRatio: 4.0 / 3.0,
    size: { width: 375, height: 667 },
  }

  constructor (props: Props) {
    super(props)
    this.fetchImageSize(this.previewImageURL(props))
  }

  previewImageURL (props: Props): ?string {
    if (!(props.submission.attachments && props.submission.attachments.length > 0)) {
      return null
    }

    return props.submission.attachments[0].url
  }

  fetchImageSize = (url: ?string) => {
    if (url) {
      Image.getSize(url, this.imageSizeLoaded)
    }
  }

  imageSizeLoaded = (width: number, height: number) => {
    this.setState({ aspectRatio: width / height })
  }

  onScrollViewLayout = ({ nativeEvent }: { nativeEvent: { layout: { width: number, height: number }}}) => {
    const { width, height } = nativeEvent.layout
    this.setState({ size: { width, height } })
  }

  onFollowLink = () => {
    if (this.props.submission.url) {
      Linking.openURL(this.props.submission.url)
    }
  }

  UNSAFE_componentWillReceiveProps (newProps: Props) {
    const newURL = this.previewImageURL(newProps)
    if (newURL && newURL !== this.previewImageURL(this.props)) {
      this.fetchImageSize(newURL)
    }
  }

  render () {
    const previewUnavailable = i18n('Preview Unavailable')
    const submissionExplanation = i18n('This submission is a URL to an external page. We\'ve included a snapshot of what the page looked like when it was submitted.')
    const preview = i18n('URL Preview Image')

    const { submission } = this.props
    const imageHeight = this.state.size.width / this.state.aspectRatio

    const image = submission.attachments && submission.attachments.length > 0
      ? <ScrollView
        onLayout={this.onScrollViewLayout}
        contentContainerStyle={styles.scrollView}
        maximumZoomScale={2.0}
        minimumZoomScale={0.5}
        showsHorizontalScrollIndicator
        showsVerticalScrollIndicator
      >
        <Image
          accessible
          accessibilityLabel={preview}
          testID='url-submission-viewer.preview'
          style={[styles.image, { height: imageHeight }]}
          source={{ uri: submission.attachments[0].url }}
        />
      </ScrollView>
      : <Paragraph style={{ alignSelf: 'center' }}>
        {previewUnavailable}
      </Paragraph>

    return (
      <ScrollView
        contentContainerStyle={{ flex: 1 }}
        contentInset={{ bottom: this.props.drawerInset }}
      >
        <Text
          testID='url-submission-viewer.explanation'
        >
          {submissionExplanation}
        </Text>
        <LinkButton
          testID='url-submission-viewer.url'
          style={styles.linkButton}
          textStyle={styles.linkButtonText}
          onPress={this.onFollowLink}
        >
          { submission.url || '' }
        </LinkButton>
        {image}
      </ScrollView>
    )
  }
}

const styles = StyleSheet.create({
  scrollView: {
    alignItems: 'center',
    justifyContent: 'flex-start',
  },
  image: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
  },
  linkButton: {
    alignSelf: 'center',
    paddingVertical: 12,
  },
  linkButtonText: {
    fontSize: 18,
  },
})
