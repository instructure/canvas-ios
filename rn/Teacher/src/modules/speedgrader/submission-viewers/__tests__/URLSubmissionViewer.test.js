//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import { Image, Linking } from 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import URLSubmissionViewer from '../URLSubmissionViewer'
import setProps from '../../../../../test/helpers/setProps'
import explore from '../../../../../test/helpers/explore'

const t = {
  ...require('../../../../__templates__/submissions.js'),
  ...require('../../../../__templates__/attachment.js'),
}

Image.getSize = jest.fn()
const mockOpenURL = jest.fn()
jest.mock('react-native/Libraries/Linking/Linking', () => ({ openURL: mockOpenURL }))
jest.mock('react-native/Libraries/Components/Touchable/TouchableOpacity', () => 'TouchableOpacity')

const submission = t.submission({
  attachment: t.attachment(),
  submission_type: 'online_url',
  url: 'https://my-blog-about-rockets.com/2017/6/1/rockets-are-splosive',
})

test('url submission renders correctly', () => {
  const tree = renderer.create(
    <URLSubmissionViewer submission={submission} drawerInset={0} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('on new url received, fetch image height if attachment', () => {
  const rendered = renderer.create(
    <URLSubmissionViewer submission={submission} drawerInset={0} />
  )
  const fetch = jest.fn()
  rendered.getInstance().fetchImageSize = fetch

  setProps(rendered, { submission: { ...submission, attachments: [] } })
  expect(fetch).not.toHaveBeenCalled()

  const newPost = 'https://my-blog-about-rockets.com/2017/6/2/liftoff'
  const oldAttachments = submission.attachments || []
  const updated = {
    ...submission,
    attachments: [{ ...oldAttachments[0], url: newPost }],
  }
  setProps(rendered, { submission: updated })
  expect(fetch).toHaveBeenCalledWith(newPost)
})

test('on scrollview layout, the state is updated', () => {
  const rendered = renderer.create(
    <URLSubmissionViewer submission={submission} drawerInset={0} />
  )

  rendered.getInstance().onScrollViewLayout(
    { nativeEvent: { layout: { width: 200, height: 200 } } }
  )

  expect(rendered.getInstance().state.size).toEqual({ width: 200, height: 200 })
})

test('loaded image sets aspect ration for rendered image view', () => {
  const rendered = renderer.create(
    <URLSubmissionViewer submission={submission} drawerInset={0} />
  )

  rendered.getInstance().imageSizeLoaded(800, 600)

  expect(rendered.getInstance().state.aspectRatio).toEqual(800 / 600)
})

test('changing the fetches the dimensions', () => {
  const url = 'https://my-blog-about-rockets.com/2017/6/2/propulsion_ftw'
  const attachments = submission.attachments || [t.attachment()]
  const newSub = {
    ...submission,
    attachments: [{ ...attachments[0], url }],
  }

  const rendered = renderer.create(
    <URLSubmissionViewer submission={submission} drawerInset={0} />
  )

  setProps(rendered, { submission: newSub })
  expect(Image.getSize).toHaveBeenCalledWith(url, rendered.getInstance().imageSizeLoaded)
})

test('tapping the url opens the link in Safari', () => {
  const tree = renderer.create(
    <URLSubmissionViewer submission={submission} drawerInset={0} />
  ).toJSON()

  const urlButton = explore(tree).selectByID('url-submission-viewer.url')
  urlButton && urlButton.props.onPress()

  expect(Linking.openURL).toHaveBeenCalledWith(submission.url)
})
