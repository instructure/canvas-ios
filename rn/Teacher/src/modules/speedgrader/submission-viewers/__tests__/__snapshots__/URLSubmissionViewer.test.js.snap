// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`url submission renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flex": 1,
    }
  }
  contentInset={
    Object {
      "bottom": 0,
    }
  }
>
  <View>
    <Text
      testID="url-submission-viewer.explanation"
    >
      This submission is a URL to an external page. We've included a snapshot of what the page looked like when it was submitted.
    </Text>
    <TouchableOpacity
      accessibilityRole="button"
      hitSlop={
        Object {
          "bottom": 15,
          "left": 15,
          "right": 15,
          "top": 15,
        }
      }
      onPress={[Function]}
      style={
        Object {
          "alignSelf": "center",
          "paddingVertical": 12,
        }
      }
      testID="url-submission-viewer.url"
    >
      <View>
        <Text
          style={
            Array [
              Object {
                "color": "#2D3B45",
                "fontSize": 16,
              },
              Array [
                Object {
                  "color": "#008EE2",
                  "fontSize": 16,
                  "fontWeight": "500",
                },
                Object {
                  "fontSize": 18,
                },
              ],
            ]
          }
        >
          https://my-blog-about-rockets.com/2017/6/1/rockets-are-splosive
        </Text>
      </View>
    </TouchableOpacity>
    <Text
      style={
        Array [
          Object {
            "color": "#8B969E",
            "fontSize": 16,
            "lineHeight": 23,
          },
          Object {
            "alignSelf": "center",
          },
        ]
      }
    >
      Preview Unavailable
    </Text>
  </View>
</RCTScrollView>
`;
